### ISSUE TITLE
Add custom public data sources

### ISSUE THREAD

**Issue by oduludo** [oduludo/eol#5]
We currently use https://endoflife.date/ as datasource, but that does not contain all data. We have to come up with an easy way to serve additional custom data sources. This should be as easy as possible to host and the CLI application should be configurable to make use of these custom sources.

An idea would be to have a JSON file in your/a repository (which must be accessible by the CLI application though... so private repos may get support only later) following a structure similar to the cycle overview endpoint (e.g. https://endoflife.date/api/amazon-linux.json).

The CLI application has to accept a flag to apply the additional datasource, e.g.: `eol check eol-cli 0.1.0 --source=https://github.com/oduludo/eol`. This should fetch a file located on https://github.com/oduludo/eol/eol-cli.json which contains a JSON array with one or more lifecycles.

---
Adding support for private repos can be done later and is out of scope for now.

**Comment by oduludo** [comment]
> Adding support for private repos can be done later and is out of scope for now.

Not to support private repos, but if you don't want your private software's versioning to be publicly visible and still use the EOL tool, support can be added for encrypting/decrypting the versioning file that's publicly available.

EOL would allow the JSON file to be encrypted using a key. The encrypted file can be published and maintained (through decrypting, updating and encrypting with the same key again) via e.g. a public repo. By adding the decryption key to you CI setup, you can decrypt the encrypted JSON file for use as a datasource to check the EOL of your custom software. The data is essentially 'hiding in plain sight' this way.

**Comment by oduludo** [comment]
Custom datasources would ideally be configured on the machine only once, to then be called when invoking `eol` commands. The alternative would be to provide the datasource and optionally a decryption key each time it has to be checked. Storing these config values can be done in `$HOME/.config/eol-cli` (on Linux and MacOS).

**Comment by oduludo** [comment]
This feature hooks into the `check` command.

Some implementation details:
- Custom datasource is set via the `--source` flag, e.g. `--source=<URL>`.
  - Multiple URLs may be configured, separated by commas (`,`): `--source=<URL1>,<URL2>`.
- By default the configured additional datasources are added on top of the endoflife.date datasource (the 'root' source).
  - To exclusively use the custom datasources, configure them with the `--xsource` flag instead of the `--source` flag.
- Custom datasources take precedence over the root source.
  - If the requested `resource@version` is present in both the root source and custom datasources, the data from the custom source is used.
  - If the requested `resource` is in the custom datasource, but the particular `version` is not, check on the root datasource.
  - If a `resource@version` is in multiple of the configured custom datasources, the order of the custom datasource URLs defines what datasource is used.
- When working with an encrypted datasource, the decryption key is passed via the `--key` flag. E.g. `--key=foobar123`.
  - If multiple encrypted datasources are used, multiple keys can be configured similar to the datasource URLs. The order of the keys has to match the order of the URLs. E.g. `--source=<URL1>,<URL2> --key=<KEY_URL1>,<KEY_URL2>`.
  - If some datasources are encrypted and some are not, `_` can be used for a key placeholder. E.g. `--source=<URL1>,<URL2> --key=_,<KEY_URL2>` if only URL2 is an encrypted source.
  - The number of URLs and keys has to match. If there's a mismatch, the program will exit.

---

Datasources are JSON files, which may be encrypted (obfuscated). The aim of `eol` when encrypting is not to come up with something unbreachable, but to obfuscate the data enough so it does not make sense to anyone who randomly encounters it. Some tools are added to `eol` to easily do this, all located under the `source` command (`eol source <any of the commands below>`).

To check whether a JSON file's structure is valid to be used as a datasource, the `verify-source` command is available. This format has yet to be defined. `verify-source` takes a `--url` flag as well as an optional `--key` flag for encrypted files.

To encrypt a datasource JSON file, the `encrypt-source` command is available. This takes a `--file` flag which will point the command to the file to encrypt. Optionally a `--key` flag may be added to use that particular key to encrypt the source. This is useful when updating an existing file. It can then be encrypted with the same key already in use by pipelines, project members etc. If no key is explicitly provided, the command will generate a key and output it. It has to be stored somewhere safe as it won't be shown again. If the `--to` flag is set, the encrypted file will be written to that path. Otherwise it will just overwrite the file on the `--file` location.

To decrypt a datasource to a JSON file, the `decrypt-source` command is available. This takes a `--from` flag which will point to the source (has to be a local file) and optionally a `--to` file to write the encrypted data to. If `--to` is not set, the `--from` file will be overwritten. The `--key` flag is required, as it holds the key to actually decrypt the data.

---

In addition to the `check` command being able to use custom datasources, the `list-versions` and `list-resources` commands offer similar functionality. The flags remain the same. As there is no checking to do, precedence is irrelevant here. Resources and versions will be listed per datasource.

---

Configuration and usage of persistent datasources is yet to be designed.

**Comment by oduludo** [comment]
[Update]

This has been implemented. All of the features discussed above have been implemented, except for:
- Using datasources from private repos
- Persistence of datasources
  - I think this is fairly useless, as the aim of EOL is to be used primarily in CI/CD workflows. There is no point in persisting data and configurations for a VM/environment that will be destroyed once its job has finished.
